### PR TITLE
Normalize sql statements to elide literal numbers and strings.

### PR DESCRIPTION
**Pull request by johnbley**
Normalize sql statements using a javacc-based lexer.

- Replace sql numeric and string literals with a ?
- Unit tests for sql normalization, including edge cases like thousands of quote marks
- Adjust existing instrumentation tests accordingly to expect normalized sql in spans
- Add configuration option to disable this feature
- Fix weird edge case in muzzle: ReferenceMatcher's handling of primitive values was preventing helper classes from using primitive fields (primitive arrays were fine) due to two different reflective type systems in use with different representations of said primitive types.

### CODE REVIEW

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/config/Config.java [hunk 1]
```diff
@@ -96,6 +96,9 @@
   private static final String DEFAULT_TRACE_EXECUTORS = "";
   private static final String DEFAULT_TRACE_METHODS = null;
 
+  public static final String DISABLE_SQL_NORMALIZER = "disable.sql.normalizer";
+  public static final boolean DEFAULT_DISABLE_SQL_NORMALIZER = false;
```
Comment (trask) [inline agent-bootstrap/src/main/java/io/opentelemetry/auto/config/Config.java:100]: I hear you about naming `boolean` settings so that the default maps to `false`, but I think this naming would be more consistent with the existing settings:

```suggestion
  public static final String SQL_NORMALIZER_ENABLED = "sql.normalizer.enabled";
  public static final boolean DEFAULT_SQL_NORMALIZER_ENABLED = true;
```

Comment (johnbley) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/config/Config.java:100]: Agreed, will turn into another commit to fix the test though 😄

**Comment by johnbley** [comment]
> I think it would be nice for the normalization to happen off the main thread (in a span processor pipeline, prior to sending to exporter), but we can revisit that later (need to address #368 first 😄).

It's pretty darn fast (sub-sub-millisecond for normal sql), and could be made faster if need be.  But I agree, we're going to encounter things like this that necessitate some off-thread processing.

**Comment by johnbley** [comment]
Can any gradle/circleci expert on here help with the spotless license check here?  The issue is that javacc generates some files in build/generated/, and the `check` target is complaining that they don't have a license header.  I specifically saw this problem on my local builds and added an exclude for **/generated/** to the top-level `gradle/spotless.gradle` config.  This works for my builds but apparently not for circleci ones?

**Comment by trask** [comment]
I can confirm. Works locally. Fails on CI. Really not sure what's the difference. Looked a bit at the spotless gradle plugin source code even 😭.

Some ideas worth trying (in this order of invasiveness I think):

* Update `com.diffplug.gradle.spotless` to the latest `3.28.1` in `trace-java.gradle`
* Change `targetExclude` from `**/generated/**` to `build/generated/**`
* Change `targetExclude` again, this time to `build/generated/**/*.java`
* Remove `targetExclude` altogether and add `target 'src/**/*.java'`

I think the last one should work, it's not ideal as it bypasses spotless' built-in detection of Java source sets, but if the other ideas don't work, then I'm all in on the last one 👍.

**Comment by trask** [comment]
`check` finally succeeded! 😆🎉

**Comment by johnbley** [comment]
> `check` finally succeeded! 😆🎉

Yes.  I don't know if you saw the commit comment, but I did manually make sure that a missing license header in src/* does in fact still fail the build.
